feat(e2e): script and CI job to use AI for e2e test selection

cmd-ob opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This PR introduces an E2E test selection system that uses AI to analyse files changed and automatically run only the most relevant smoke tests, reducing CI time.

## How It Works

1. **AI analyses** your code changes and selects appropriate test tags
2. **Results posted** back to the PR with detailed analysis
3. **Test Matrix** are included in the result which are compatible with our current GH workflows for running E2E tests

## Key Features

- 🤖 **AI-powered selection** - Claude AI analyses diffs and recommends test tags
- 🎯 **Risk assessment** - Categorises changes as low/medium/high risk

## Usage

**For PRs:**
AI Analysis is now integrated into main CI and will run on every pull request with recommendations. (Actual test running will be in a different PR)

**For local testing:**
```bash
yarn ai-e2e --verbose  # See AI reasoning
yarn ai-e2e --output json  # Machine-readable
```

See `ai-e2e-testing.md` for complete details.

## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry:

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.



<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds Claude-driven E2E test tag selection with a CI analysis job that outputs a test matrix, plus supporting scripts, docs, and dependencies.
> 
> - **CI / GitHub Actions**:
>   - **Composite Action**: Introduces `.github/actions/ai-e2e-analysis` to run AI analysis, set outputs (`test_matrix`), and optionally comment on PRs.
>   - **Workflow**: Adds `ai-e2e-analysis` job in `.github/workflows/ci.yml` consuming `needs_e2e_build` changed files and exposing `test-matrix` output.
> - **E2E Tooling**:
>   - **AI Tag Selector**: Adds `e2e/scripts/ai-e2e-tags-selector.ts` using `@anthropic-ai/sdk` with `esbuild-register` to choose smoke `tags`, count test files, and recommend splits with JSON output.
>   - **Runner Script**: Adds `.github/scripts/ai-e2e-analysis.mjs` to execute the selector, parse results, and write GitHub Action outputs/summary.
> - **Docs**:
>   - Adds `e2e/scripts/ai-e2e-testing.md` explaining usage and options.
> - **Package / Config**:
>   - Updates `package.json` with `ai-e2e` script and dev deps `@anthropic-ai/sdk`, `esbuild`, `esbuild-register`; updates `lavamoat.allowScripts` and resolutions accordingly.
>   - Updates `.depcheckrc.yml` ignores and `.eslintignore` for new files.
>   - Exports `tags` from `e2e/tags.js`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit b4953f11d96754f4fa2f6cb9b13cd7edbc881527. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

cmd-ob committed:
feat: Implement AI-powered E2E test selection and reporting workflow

- Added reusable GitHub Actions workflow for reporting E2E test results.
- Created documentation for the AI E2E Testing system, detailing architecture and usage.
- Introduced AI E2E Tag Selector script to analyze code changes and select relevant tests.
- Updated E2E tags module to export tags for AI analysis.
- Added README for smart E2E test selection with local and CI usage instructions.

cmd-ob committed:
feat: Add AI E2E test script and update dependencies for esbuild and anthropic AI SDK

cmd-ob commented:
@SocketSecurity ignore @anthropic-ai/sdk

cmd-ob committed:
allow script fix

cmd-ob committed:
fix: Update .depcheckrc.yml to ignore esbuild and esbuild-register for AI E2E script compilation

cmd-ob committed:
Merge branch 'main' into MMQA-991-clean

racitores reviewed (commented):
Nice to see AI stuff in the pipelines

cmd-ob committed:
feat: add AI E2E analysis script and integrate with workflow

cmd-ob commented on .github/workflows/ai-e2e-selection.yml:
```diff
@@ -0,0 +1,645 @@
+        env:
+          E2E_CLAUDE_API_KEY: ${{ secrets.E2E_CLAUDE_API_KEY }}
+        run: |
+          echo "🤖 Running AI analysis..."
```
updated

ngalee8311 commented:
> ## 🔍 AI E2E Analysis Report
> 
> **Risk Level:** high | **Selected Tags:** SmokeAccounts, SmokeConfirmations, SmokeConfirmationsRedesigned, SmokeIdentity, SmokeNetworkAbstractions, SmokeNetworkExpansion, SmokeTrade, SmokeWalletPlatform
> 
> **🤖 AI Analysis:**
> > Dependency changes (yarn.lock/package.json) detected which typically warrant running ALL smoke test tags due to potential wide-reaching effects. Additionally, changes span multiple critical areas: core Engine components, rewards controller, confirmations flow, account management, network handling, ramp functionality, and E2E test infrastructure. The breadth and depth of changes across 280 files including core infrastructure makes comprehensive testing essential.
> 
> **📊 Analysis Results:**
> - **Mode:** Analysis-Only ✅ (no builds or tests)
> - **Confidence:** 95%
> 
> _🔍 [View complete analysis](https://github.com/MetaMask/metamask-mobile/actions/runs/18034521199) • Analysis completed successfully_
> 
> <!-- ai-e2e-analysis -->
>

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into MMQA-991-clean

cmd-ob committed:
refactor: replace bash AI analysis script with Node.js implementation

cmd-ob committed:
refactor: simplify AI analysis script by removing unused parameters and updating workflow triggers

cmd-ob committed:
feat: add AI E2E analysis action and integrate into CI workflow

cmd-ob committed:
feat: add checkout step to AI E2E Analysis job in CI workflow

cmd-ob committed:
feat: add permissions for AI E2E Analysis job in CI workflow

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into MMQA-991-clean

cmd-ob committed:
docs: update AI E2E test selector documentation with detailed usage and command options

cmd-ob committed:
feat: update AI E2E analysis to include main changes option in command

cmd-ob committed:
feat: add support for providing changed files as an argument to skip git commands

racitores reviewed (dismissed):
Looks fantastic. Added a couple of things that is up to you to do it. It can be always done  in future PRs

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into MMQA-991-clean

cmd-ob commented on .github/actions/ai-e2e-analysis/action.yml:
```diff
@@ -0,0 +1,157 @@
+name: 'AI E2E Analysis'
+description: 'Run AI-powered E2E test selection analysis based on code changes'
+inputs:
+  changed-files:
+    description: 'Pre-computed list of changed files (optional - if not provided, will compute via git)'
+    required: false
```
skip analysis if changed files is []

cmd-ob commented on scripts/e2e/README.md:
move to e2e folder

cmd-ob committed:
fix: Update AI E2E Analysis action to require changed files and simplify command usage

cmd-ob committed:
Changes from PR comments

- Moved related files under /e2e
- Updated the package.json to correct the path for the AI E2E script.
- Removed the outdated README file for the local AI E2E test selector script.

cmd-ob committed:
Merge branch 'main' of github.com:MetaMask/metamask-mobile into MMQA-991-clean

cmd-ob committed:
feat: AI E2E Analysis to run only on pull_request events with changed files

cmd-ob committed:
fix: streamline AI E2E analysis by removing redundant checkout step

cmd-ob committed:
fix: include .github/scripts in sparse-checkout for AI E2E analysis

cmd-ob committed:
fix: include additional directories in sparse-checkout for AI E2E analysis

cmd-ob committed:
Merge branch 'main' into MMQA-991-clean

jvbriones commented on .github/workflows/ci.yml:
```diff
@@ -226,7 +226,37 @@ jobs:
     uses: ./.github/workflows/needs-e2e-build.yml
-      
+
+  ai-e2e-analysis:
```
can we make this job to always pass for now?

cmd-ob committed:
add continue-on-error for ai-e2e-analysis job

cmd-ob committed:
refactor: reduce readme

cmd-ob committed:
refactor: streamline selection guidelines and remove dependency change warnings

cmd-ob committed:
Merge branch 'main' into MMQA-991-clean

Qbandev reviewed (commented):
Why is this file `e2e/scripts/ai-e2e-testing.md` in the script folder and not in the docs?
